### ISSUE TITLE
Proposal: A way to solve the strange behaviour with Xcode's test navigator

### ISSUE THREAD

**Issue by sharplet** [Quick/Quick#439]
I found a _whole bunch_ of potentially related issues:

- https://github.com/Quick/Quick/issues/14
- https://github.com/Quick/Quick/issues/157
- https://github.com/Quick/Quick/issues/221
- https://github.com/Quick/Quick/issues/219
- https://github.com/Quick/Quick/issues/373
- https://github.com/Quick/Quick/issues/427
- https://github.com/Quick/Quick/issues/429

I did a bit of digging into what might be going on here.

When you press Cmd-U, XCTest seems to do roughly the following:

- For each bundle, builds the default test suite for that bundle
- For each bundle, creates a test suite named "All tests"
- For each test case class in the bundle, creates a test suite for that class by sending `+[XCTestSuite testSuiteForTestCaseClass:]`, and adds it to the "All tests" test suite
- Internally that method then sends `+testInvocations`, which is where Quick kicks in and builds the invocations from its example suite

When you click on a test suite in the test navigator, the behaviour is different:

- Only your test bundle is invoked
- A test suite named "Selected tests" is created
- XCTest sends `+[XCTestSuite testSuiteForTestCaseClassWithName:]`, where the name argument looks like this: `<spec-class>/<selector>`, e.g., `NewSpec/stuff__is_cool_pathtoNewSpecswift_6()`. (The trailing `()` might not be there for Objective-C, I haven't tested that.) **It sends this for each selector in the test navigator.** These test suites are added to the "Selected tests" test suite.
- `+testInvocations` is never sent.
- This method in turn sends the spec class the `+testCaseWithSelector:` message. It appears do do this once for each specific test invocation selector listed in the test navigator.

---

Here's the problem:

- You've run the tests once, and you can see everything in the test navigator.
- You run a specific spec, and everything looks fine.
- You change the description of one of the example group nodes, either a `describe()`, `context()` or `it()`.
- This changes the selector that will be generated at runtime. This is a different selector to the one that Xcode has cached in the test navigator.
- You re-run just that spec.
- When XCTest sends the test case class the `+testCaseWithSelector:` message, the `selector` that gets passed _doesn't exist anymore_.
- You get the weird `-[MyTests.MySpec (null)]` message.
- Stuff may even get weirder.

Semantically, when you try and run a specific spec, you really want to run all the examples in that spec. However, instead of asking our spec class to re-run all its examples, it instead is specifying a finite set of the test case methods it already knows about.

The only thing I've been able to come up with so far to fix this is to swizzle `+[XCTestSuite testSuiteForTestCaseWithName:]` and override it with our own semantics. But the selector we're asked for _might not even exist_.

So that means that we'd probably have to go up a level:

- Swizzle `-[XCTestSuite initWithName:]`
- Look for a test suite being created with the specific name "Selected tests"
- Swizzle this instance's class to be an instance of our own `XCTestSuite` subclass
- Build our own "Selected tests" test suite, and add it.
- Ignore any other tests that Xcode tries to add to this suite.

---

There are other factors to consider, and a risk is that we get even more coupled to Xcode. This implementation would be a start, but it's not ideal.

And with a version of XCTest going open-source, perhaps it would be better to adopt a wait-and-see approach?

**Comment by modocache** [comment]
Awesome investigation, thanks @sharplet!! :bouquet: 

> And with a version of XCTest going open-source, perhaps it would be better to adopt a wait-and-see approach?

I think it's going to be a long time before [apple/swift-corelibs-xctest](https://github.com/apple/swift-corelibs-xctest) gets anywhere close to interacting with Xcode. I think we can solve this issue in parallel with #433. I envision `QuickSpec.{h,m}` acting as an integration layer with Apple's closed-source XCTest, and any swizzling logic can probably live alongside of it.

The swizzling approach seems solid. That and a radar should help a lot of people out.

**Comment by PaulTaykalo** [comment]
I'll add some information here. It won't be able to fix everything by just implementing something in tests library. Well, I would see that I don't see currently any easy way to do this:

Here's a catch:

From time to time Xcode runs his own Indexer, which search for tests Locations in the openede Xcode files
```
#0	0x0000000100fd507c in -[IDETest setIndexLocation:] ()
#1	0x00007fff9f3abb41 in -[NSObject(NSKeyValueObservingPrivate) _changeValueForKey:key:key:usingBlock:] ()
#2	0x00007fff9f407c91 in _NSSetObjectValueAndNotify ()
#3	0x0000000100fd4a46 in -[IDETest _primitiveSetIndexLocation:] ()
#4	0x0000000100fd46e1 in -[IDETest _setIndexLocation:forSubtest:] ()
#5	0x0000000100fd4b50 in -[IDETest _setIndexLocation:] ()
#6	0x000000010c736c5e in -[Xcode3OCUnitTestable updateExistingTests:toMatchNewIdentifiersAndLocations:] ()
#7	0x000000010c736ba1 in -[Xcode3OCUnitTestable updateExistingTests:toMatchNewIdentifiersAndLocations:] ()
#8	0x000000010c7366d7 in __39-[Xcode3OCUnitTestable initWithTarget:]_block_invoke ()
#9	0x000000010c610773 in __85-[Xcode3OCUnitTestableDataSource _setIndexableIdsToTestableIdsToDefinitionLocations:]_block_invoke ()
#10	0x00007fff8e9d2a6f in __51-[__NSSetM enumerateObjectsWithOptions:usingBlock:]_block_invoke ()
#11	0x00007fff8e9d2979 in -[__NSSetM enumerateObjectsWithOptions:usingBlock:] ()
#12	0x000000010c610593 in -[Xcode3OCUnitTestableDataSource _setIndexableIdsToTestableIdsToDefinitionLocations:] ()
#13	0x000000010c60fbfa in __49-[Xcode3OCUnitTestableDataSource _searchForTests]_block_invoke229 ()
```

And since this datasource doesn't know what it should be seaching for in terms of QuickTests/KiwiTests, then it only finds tests in `- (void)test.*` methods.

After this happens, there's no more information about runtime generated tests we have in Kiwi/Quick

**Comment by modocache** [comment]
Thanks, @PaulTaykalo! Filing radars may be the best approach to this problem. Post links here if you've submitted any.

> It won't be able to fix everything by just implementing something in tests library.

I'm mostly concerned with tests not being run. I'm fine if Xcode is unable to detect tests properly (for now). I think @sharplet's solution may be able to at least ensure everyone's tests *run*, even if they dont' show up properly in Xcode's test navigator.

**Comment by PaulTaykalo** [comment]
Here's the method, which is actually called by indexer
https://github.com/luisobo/Xcode-RuntimeHeaders/blob/master/IDEFoundation/IDEIndexDatabaseQueryProvider.h#L120

**Comment by PaulTaykalo** [comment]
@modocache How *radar* should help? 
It's not technically a bug :) It's just how it works.

**Comment by modocache** [comment]
Kiwi, Specta, and Quick all use the `+[XCTestCase testInvocations]` method, the only officially supported way of dynamically defining tests to run.

The fact that Xcode drops tests on the floor is a bug, in that it means the public invocations API cannot be used reliably.

The fact that Xcode does not provide any API for people using the `+testInvocations` API to display tests in the navigator is a feature request that could also be made via a radar.

**Comment by brentleyjones** [comment]
I can't provide much help in fixing this, but I would just like to chime in and say this is _the_ reason why I currently only use Nimble and not Quick. I love Spec style testing, but it's really hard to do Red, Green, Refactor when I add a test, tell Xcode to re-run the spec, and it doesn't run the new test (only Cmd-U does). I can even live without the gutter buttons, though that is also pretty annoying as well.

**Comment by modocache** [comment]
@brentleyjones I hear ya. :disappointed:

I think the one resource we all lack is time. Besides contributing a pull request with a fix (which really would be ideal), I think there are a few ways to help:

- File well-written radars for this bug. Attach Quick as part of the sample project and explain the issues around test navigator discovery and dropped tests.
- Investigating workarounds. Even if they're inconvenient for end users it'd be helpful to be able to respond to issues with an interim solution.

**Comment by modocache** [comment]
@sharplet I'd love to see this happen. I think coupling to Xcode on OS X is fine for now; the core libraries like swift-corelibs-xctest and swift-corelibs-foundation are coupling their build environments to Xcode right now as well. We can think of Linux support (which of course means no Xcode) later, as part of #433.

Do you have something on a local branch somewhere you can push up?

**Comment by sharplet** [comment]
I have something mostly working, I'll try push it up in the next day or so. 🙃

**Comment by levous** [comment]
@sharplet thank you!  Can't wait.  I'd begrudgingly returned to using xctest directly though I truly love Quick.  My team was not writing or running tests because of these Xcode issues.  I'm excited to try again

**Comment by sharplet** [comment]
Sorry about the delay here everyone!

I was working on this in a submodule of another project, and I accidentally blew away the submodule. 🙃

Shouldn't take very long to rewrite and I'll have a branch for people to test against.

**Comment by sharplet** [comment]
I've just opened #452 which adds a fix for this behaviour. Please test this out in your own projects and let me know what you think!

**Comment by modocache** [comment]
I've merged #452, which I'm hoping resolves this issue for all of you. Many thanks to @sharplet.

Keep in mind what has been fixed: all tests should now be run, every time. However, not all tests are always displayed in the Xcode UI. 

I'll open another issue for that bug. In the meantime, this is a great step forward! I think we can all agree that the tests running as expected is of paramount importance, and the UI comes second to that.

**Comment by Dschee** [comment]
Thanks anyone for investigating and trying to solve this weird issue with XCode 7.2. I just came across this, too and I don't know if it's coincidence or not but in two of my specs both the last `describe` block wasn't shown in the test navigator so I just tried adding an empty block like this:

``` Swift
        // This line is only needed to see failed line markers on XCode 7.2 if tests fail in some tests
        describe("XCode 7.2 bug") {
            it("workaround") {
                expect(true).to(beTrue())
            }
        }
```

This actually solved the issue for my specs. I also tried changing the order of my describe blocks and added an `expect(true).to(beFalse())` to each of my describe blocks so I could see which tests were actually run and which shown. All tests were always run (the final "Tests failed" appeared all the time) and it was always the last describe block that didn't appear in the test navigator.

Again, this could be a coincidence in my specific project but maybe this helps investigating the issue with the test navigator. Is there an issue opened for that btw?

**Comment by Dschee** [comment]
I just set up an example project to reproduce the issue and it worked. Here's the project where you can see and investigate the issue yourself:
https://github.com/Dschee/WeirdTestNavigatorDemo

It's possible that this only happens on my XCode 7.2 installation though, so it would be great to hear if the project behaves the same for you guys. On my machine both describe blocks fail but only one appears in the project navigator (the first). When I add another block the first two are shown etc.

**Comment by Dschee** [comment]
I just further investigated the weird behavior by using the linked project and it seems like the last it block when sorting all tests by name doesn't show up. That explains the following example test result for me:

<img width="1002" alt="bildschirmfoto 2016-01-23 um 21 25 00" src="https://cloud.githubusercontent.com/assets/6942160/12532579/d71a88f6-c217-11e5-8858-a27ef814955b.png">

**Comment by timbroder** [comment]
Thanks @Dschee so far this is working for me. Will report back if that changes

@modocache Did you open one or should we use https://github.com/Quick/Quick/issues/427?

> I'll open another issue for that bug.

**Comment by programmerdave** [comment]
@modocache Just pinging you again for what @timbroder asked you. Should we be following this issue or did you open another one to fix the UI bug?
